orbstation/code/map/automapper/area_spawn_subsystem.dm

```diff
@@ -0,0 +1,376 @@
+
+				if(
+					!is_type_in_list(found_movable, allowed_diagonal_objects_list) \
+					&& (found_movable.density || is_type_in_list(found_movable, restricted_objects_list))
+				)
+					num_diagonal_objects++
+					break
+
+	switch(mode)
+		if(AREA_SPAWN_MODE_OPEN)
+			// For non-wall hug
+			// #1 priority is totally empty
+			// #2 priority is being in the middle of the room
+			return (totally_empty ? 10 : 0) + (4 - num_dense_found)
+
+		if(AREA_SPAWN_MODE_HUG_WALL)
+			// For wall hugging, must be against wall, and not touching another dense object as it may completely block it.
+			if(num_walls_found == 0 || found_dense_object || num_walls_found == 4)
+				return 0
+
+			// #1 Priority after that: be in a totally empty square
+			// #2 (marginally) have clear diagnals
+			// #3 favor being in a cozy wall nook
+			// #4 be in a big room/hallway so we don't pinch a room down to 1 square of passage.
+			return (totally_empty ? 1000 : 0) + (400 - num_diagonal_objects * 100) + (num_walls_found * 10) + num_very_open_floors
+
+		if(AREA_SPAWN_MODE_MOUNT_WALL)
+			// For mounting to walls. Must be against wall.
+			if(num_walls_found == 0 || num_walls_found == 4)
+				return 0
+
+			// #1 Priority after that: be in a totally empty square
+			// #2, actually don't be in a nook!
+			return (totally_empty ? 10 : 0) + (4 - num_walls_found)
+
+	CRASH("Invalid area spawn mode [mode]!")
+
+/**
+ * Pick a turf candidate and remove from the list.
+ *
+ * Only picks one of the highest priority ones.
+ *
+ * Arguments:
+ * * turf_candidates - Turf candidate list produced by
+ */
+/datum/controller/subsystem/area_spawn/proc/pick_turf_candidate(list/list/turf/turf_candidates)
+	// Pick-n-take highest priority.
+	var/list/turf/sublist = turf_candidates[peek(turf_candidates)]
+	var/turf/winner = pick_n_take(sublist)
+
+	// To be safe, remove the neighbors too.
+	for(var/dir in GLOB.cardinals)
+		var/turf/neighbor = get_step(winner, dir)
+		sublist -= neighbor
+
+	// Remove this priority if it's now empty.
+	if(!LAZYLEN(sublist))
+		pop(turf_candidates)
+
+	// Extremely specific, but landmarks are immediately destroyed when created so can't be detected another way.
+	// This is the only landmark list that normally creates solid objects in non-maintenance spaces.
+	GLOB.secequipment -= winner
+
+	return winner
+
+/**
+ * Area spawn datums
+ *
+ * Use these to spawn atoms in areas instead of placing them on a map. It will select any available open and entering turf.
+ */
+/datum/area_spawn
+	/// The target area for us to spawn the desired atom, the list is formatted, highest priority first.
+	var/list/target_areas
+	/// The atom that we want to spawn
+	var/desired_atom
+	/// The amount we want to spawn
+	var/amount_to_spawn = 1
+	/// See orbstation/_global/automapper.dm
+	var/mode = AREA_SPAWN_MODE_OPEN
+	/// Map blacklist, this is used to determine what maps we should not spawn on.
+	var/list/blacklisted_stations = list("Biodome", "Runtime Station", "MultiZ Debug")
```

Comment:
Why Biodome?

Reply:
Yeah, it's just to discourage using this when there's literally no reason to.

Reply:
It's because this system runs on every map save for the blacklisted ones, and we don't really need it to run on Biodome since we can just add things to it ourselves.